Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm

```diff
@@ -31,7 +31,111 @@
+- (NSString *)titleForWebExtensionContext:(WKWebExtensionContext *)context { return _dictionary[@"title"]; }
+- (NSString *)urlForWebExtensionContext:(WKWebExtensionContext *)context { return _dictionary[@"url"]; }
+- (NSInteger)indexForWebExtensionContext:(WKWebExtensionContext *)context { return [_dictionary[@"index"] integerValue]; }
+- (NSArray<id<_WKWebExtensionBookmark>> *)childrenForWebExtensionContext:(WKWebExtensionContext *)context {
+    NSArray *childDictionaries = _dictionary[@"children"];
+    if (!childDictionaries)
+        return nil;
+    NSMutableArray<id<_WKWebExtensionBookmark>> *children = [NSMutableArray array];
+    for (NSDictionary *childDict in childDictionaries)
+        [children addObject:[[_MockBookmarkNode alloc] initWithDictionary:childDict]];
+    return children;
+}
+@end
+
+
+@interface TestBookmarksDelegate : NSObject <WKWebExtensionControllerDelegatePrivate>
+@property (nonatomic, strong) NSMutableArray<NSMutableDictionary *> *mockBookmarks;
+@property (nonatomic) NSInteger nextMockBookmarkId;
+@end
+
+static NSMutableDictionary *findParentInMockTree(NSMutableArray<NSMutableDictionary *> *tree, NSString *parentId)
+{
+    for (NSMutableDictionary *node in tree) {
+        if ([node[@"id"] isEqualToString:parentId])
+            return node;
+        id children = node[@"children"];
+        if (children && [children isKindOfClass:[NSMutableArray class]]) {
+            NSMutableDictionary *found = findParentInMockTree(children, parentId);
+            if (found)
+                return found;
+        }
+    }
+    return nil;
+}
+
+
+@implementation TestBookmarksDelegate
+- (instancetype)init {
+    self = [super init];
+    if (self) {
+        _mockBookmarks = [NSMutableArray array];
+        _nextMockBookmarkId = 100;
+    }
+    return self;
+}
+
+- (void)_webExtensionController:(WKWebExtensionController *)controller bookmarksForExtensionContext:(WKWebExtensionContext *)context completionHandler:(void (^)(NSArray<NSObject<_WKWebExtensionBookmark> *> *, NSError *))completionHandler
+{
+    NSMutableArray<NSObject<_WKWebExtensionBookmark> *> *results = [NSMutableArray array];
+    for (NSDictionary *bookmarkDict in self.mockBookmarks)
+        [results addObject:[[_MockBookmarkNode alloc] initWithDictionary:bookmarkDict]];
+    completionHandler(results, nil);
+}
+
+- (void)_webExtensionController:(WKWebExtensionController *)controller createBookmarkWithParentIdentifier:(NSString *)parentId index:(NSNumber *)index url:(NSString *)url title:(NSString *)title forExtensionContext:(WKWebExtensionContext *)context completionHandler:(void (^)(NSObject<_WKWebExtensionBookmark> *, NSError *))completionHandler
+{
+    NSMutableDictionary *newBookmark = [NSMutableDictionary dictionary];
+    newBookmark[@"title"] = title;
+    if (url)
+        newBookmark[@"url"] = url;
+    NSString *newId = [NSString stringWithFormat:@"%ld", (long)self.nextMockBookmarkId];
+    _nextMockBookmarkId++;
```

Comment:
Is there a reason to use `self.nextMockBookmarkId` and then `_nextMockBookmarkId++` on the next line?

Could it just be:
`NSString *newId = [NSString stringWithFormat:@"%ld", _nextMockBookmarkId++];`